### PR TITLE
Add support for private AWS ECR images

### PR DESCRIPTION
**Pull request by nickpetrovic**
Resolve BE-1553

### CODE REVIEW

**Comment by nickpetrovic** [comment]
An example that can be used with AWS IAM Identity Center (formerly AWS SSO) STS credentials.

```python
image = Image(
    base_image="111111111111.dkr.ecr.us-east-1.amazonaws.com/myapp:latest",
    base_image_creds=[
        "AWS_ACCESS_KEY_ID",
        "AWS_SECRET_ACCESS_KEY",
        "AWS_SESSION_TOKEN",
        "AWS_REGION",
    ],
)


@endpoint(image=image)
def square(i: int = 0) -> int:
    return i**2
```

File: pkg/abstractions/image/build.go [hunk 1]
```diff
@@ -41,11 +41,12 @@ type BuildOpts struct {
 	BaseImageRegistry  string
 	BaseImageName      string
 	BaseImageTag       string
+	BaseImageCreds     string
```
Comment (jsun-m) [inline pkg/abstractions/image/build.go:44]: Will all image creds always be resolved to being a single string token?

Comment (jsun-m) [inline pkg/abstractions/image/build.go:44]: Will all image creds always be resolved to being a single string token?

File: pkg/abstractions/image/build.go [hunk 2]
```diff
@@ -387,3 +367,28 @@ func (b *Builder) generatePipInstallCommand(opts *BuildOpts) string {
 	packages := strings.Join(escapedPackages, " ")
 	return fmt.Sprintf("%s -m pip install --root-user-action=ignore %s", opts.PythonVersion, packages)
 }
+
+var imageNamePattern = regexp.MustCompile(`^(?:([^/]+)(?:/))?((?:[^/:]+/)?[^/:]+)(?::([^/]+))?$`)
```
Comment (jsun-m) [inline pkg/abstractions/image/build.go:371]: Is there a source for this? We should probably add a comment to some url where this verified

File: pkg/abstractions/image/credentials.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		return "", fmt.Errorf("error decoding token: %v", err)
+	}
+
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	if len(parts) != 2 {
+		return "", fmt.Errorf("decoded token format invalid, expected username:password")
+	}
+
+	// Return the username and password
+	var username = parts[0]
```
Comment (jsun-m) [inline pkg/abstractions/image/credentials.go:85]: I suggest we use `:=` over `var` for single use local variables for most cases

**Review by jsun-m — CHANGES_REQUESTED** [review]

File: pkg/abstractions/image/build.go (hunk 2 above)
Comment (nickpetrovic) [reply pkg/abstractions/image/build.go:371]: There isn't. It was trial and error. There are tests for this that should cover all cases if there's a concern.

File: pkg/abstractions/image/build.go (hunk 1 above)
Comment (nickpetrovic) [reply pkg/abstractions/image/build.go:44]: Yes, I believe so. Skopeo only has `--src-creds` or `--username` and `--password` flags.

How we get this username/password is where the details lie e.g. from an AWS account or from a GCP account.

File: pkg/abstractions/image/credentials.go (hunk 3 above)
Comment (nickpetrovic) [reply pkg/abstractions/image/credentials.go]: Updated.

File: pkg/abstractions/image/credentials.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	if len(parts) != 2 {
+		return "", fmt.Errorf("decoded token format invalid, expected username:password")
+	}
+
+	// Return the username and password
+	username := parts[0]
+	password := parts[1]
+
+	var token = fmt.Sprintf("%s:%s", username, password)
```
Comment (jsun-m) [inline pkg/abstractions/image/credentials.go:88]: you missed one

**Review by jsun-m — APPROVED** [review]
just that last `var` and it's good.

File: pkg/abstractions/image/credentials.go (hunk 4 above)
Comment (nickpetrovic) [reply pkg/abstractions/image/credentials.go]: Updated.

**Pull request merged**